### PR TITLE
[SPARK-28077][SQL] Support ANSI SQL OVERLAY function.

### PR DESCRIPTION
**Pull request by beliefer**
## What changes were proposed in this pull request?

The `OVERLAY` function is a `ANSI` `SQL`.
For example:
```
SELECT OVERLAY('abcdef' PLACING '45' FROM 4);

SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5);

SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5 FOR 0);

SELECT OVERLAY('babosa' PLACING 'ubb' FROM 2 FOR 4);
```
The results of the above four `SQL` are:
```
abc45f
yabadaba
yabadabadoo
bubba
```

Note: If the input string is null, then the result is null too.

There are some mainstream database support the syntax.
**PostgreSQL:**
https://www.postgresql.org/docs/11/functions-string.html

**Vertica:** https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/String/OVERLAY.htm?zoom_highlight=overlay

**Oracle:**
https://docs.oracle.com/en/database/oracle/oracle-database/19/arpls/UTL_RAW.html#GUID-342E37E7-FE43-4CE1-A0E9-7DAABD000369

**DB2:**
https://www.ibm.com/support/knowledgecenter/SSGMCP_5.3.0/com.ibm.cics.rexx.doc/rexx/overlay.html

There are some show of the PR on my production environment.
```
spark-sql> SELECT OVERLAY('abcdef' PLACING '45' FROM 4);
abc45f
Time taken: 6.385 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5);
yabadaba
Time taken: 0.191 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5 FOR 0);
yabadabadoo
Time taken: 0.186 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY('babosa' PLACING 'ubb' FROM 2 FOR 4);
bubba
Time taken: 0.151 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING '45' FROM 4);
NULL
Time taken: 0.22 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING 'daba' FROM 5);
NULL
Time taken: 0.157 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING 'daba' FROM 5 FOR 0);
NULL
Time taken: 0.254 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING 'ubb' FROM 2 FOR 4);
NULL
Time taken: 0.159 seconds, Fetched 1 row(s)
```

## How was this patch tested?

Exists UT and new UT.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -454,6 +454,66 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
+    val length = ctx.freshName("length")
+    val integer = classOf[java.lang.Integer].getName
+    defineCodeGen(ctx, ev, (input, replace, pos, len) => {
+      s"""
+        boolean $isMax = $len.equals($integer.MAX_VALUE)
+        int $length = $len
+        if ($isMax) {
+          $length = $replace.toString.size
+        }
+        $input.substringSQL(1, $pos - 1) + $replace + $input.substringSQL($pos + $length, $integer.MAX_VALUE);
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]:
```
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:510: File line length exceeds 100 characters
```
Please fix the Scala style first.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: OK. I forgot it. @wangyum Thanks.

**Comment by beliefer** [comment]
Retest this please.

**Comment by beliefer** [comment]
Retest this please.

**Comment by gatorsmile** [comment]
cc @ueshin Could you take a look at this PR?

**Review by ueshin — COMMENTED** [review]
I left a comment about tests. Could you add such test cases?

Btw, sounds like we can just use `substring` and `concat` in the parser instead of adding the new `Overlay` expression. WDTY?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -454,6 +454,70 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
+      s"""
+        boolean $isMax = $len.equals($integer.MAX_VALUE);
+        int $length = $len;
+        if ($isMax) {
+          $length = $replace.toString.size;
+        }
+        UTF8String $head = $input.substringSQL(1, $pos - 1);
+        UTF8String $tail = $input.substringSQL($pos + $length, $integer.MAX_VALUE);
+        $head + $replace + $tail;
+      }
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: @ueshin Thanks for your review. I added some tests in `StringExpressionsSuite`.

[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: @ueshin I can't find any tests for string function in `DataFrameFunctionsSuite`, maybe it is used to tests `df` or array function.

[the remark this answers is not in the document]
Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: How about `StringFunctionsSuite`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -454,6 +454,67 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
+      val isMax = s"$len.equals(${Int.MaxValue})"
+      s"""if ($isMax) {
+        $length = $replace.toString().size;
+      } else {
+        $length = $len;
+      }
+      $head = $input.substringSQL(1, $pos - 1);
+      $tail = $input.substringSQL($pos + $length, ${Int.MaxValue});
+      ${ev.value} = $head + $replace + $tail;
+      """
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Btw, could you follow the string format for codes like:

```
s"""
   |...
   |...
 """.stripMargin
```

?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: @ueshin Thanks for your review. I have added some tests in `StringFunctionsSuite`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: @ueshin Thanks for your reminder. I changed to the format.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 4]
```diff
@@ -454,6 +454,68 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
+  override def nullSafeEval(inputEval: Any, replaceEval: Any, posEval: Any, lenEval: Any): Any = {
+    val inputStr = inputEval.asInstanceOf[UTF8String]
+    val replaceStr = replaceEval.asInstanceOf[UTF8String]
+    val position = posEval.asInstanceOf[Int]
+    val length = lenEval.asInstanceOf[Int]
+    Overlay.calcuate(inputStr, replaceStr, position, length)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val result = ctx.addMutableState("UTF8String", "result")
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: I don't think we need to put this to mutable state?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: OK. I removed the unreasonable mutable state.

**Comment by beliefer** [comment]
@ueshin Could you continue to review this PR?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 5]
```diff
@@ -428,6 +428,30 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     // scalastyle:on
   }
 
+  test("overlay") {
+    checkEvaluation(new Overlay(Literal("Spark SQL"), Literal("_"),
+        Literal.create(6, IntegerType)), "Spark_SQL")
```
Comment (ueshin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: nit: indent (2 spaces, not 4)

**Review by ueshin — COMMENTED** [review]
I left some comments.

Btw, I checked the Postgres' behavior and found the behavior for negative `FOR` is different:

```
postgres=# SELECT OVERLAY('abcdef' PLACING '45' FROM 4 FOR -2);
  overlay
------------
 abc45bcdef
(1 row)
```

```
spark-sql> SELECT OVERLAY('abcdef' PLACING '45' FROM 4 FOR -2);
abc45f
Time taken: 0.084 seconds, Fetched 1 row(s)
```

Do you know what the other databases' behaviors are like?
Maybe we should follow them. cc @gatorsmile

**Comment by beliefer** [comment]
@ueshin Oracle [doc](https://docs.oracle.com/en/database/oracle/oracle-database/19/arpls/UTL_RAW.html#GUID-342E37E7-FE43-4CE1-A0E9-7DAABD000369) have some description of len 
`If len is specified, it must be greater than or equal to 0. `
otherwise, throw a error.
DB2 [doc](https://www.ibm.com/support/knowledgecenter/SSGMCP_5.3.0/com.ibm.cics.rexx.doc/rexx/overlay.html) have some description of len 
`If you specify length, it must be a positive whole number or zero. The default value for length is the length of new. `
This is the difference between `PostgreSQL` and `DB2 or Oracle`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 6]
```diff
@@ -454,6 +454,66 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
   override def prettyName: String = "replace"
 }
 
+object Overlay {
+
+  def calcuate(input: UTF8String, replace: UTF8String, pos: Integer, len: Integer): UTF8String = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: OK. It's good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 7]
```diff
@@ -454,6 +454,66 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
+    val header = input.substringSQL(1, pos - 1)
+    var length = len
+    if (len < 0) {
+      length = replace.toString().length()
+    }
+    val tailer = input.substringSQL(pos + length, Int.MaxValue)
+    UTF8String.fromString(header.toString + replace.toString + tailer.toString)
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: A good idea. I learned more.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Yes, I make a typo.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 5 above)
Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: I changed it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 8]
```diff
@@ -454,6 +455,64 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
 
+object Overlay {
+
+  def calculate(input: UTF8String, replace: UTF8String, pos: Int, len: Int): UTF8String = {
+    val builder = new UTF8StringBuilder
+    builder.append(input.substringSQL(1, pos - 1))
+    builder.append(replace)
+    var length = len
+    if (len < 0) {
+      length = replace.toString().length()
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: We can use `UTF8String.numChars()`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 9]
```diff
@@ -454,6 +455,64 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
+  override def nullSafeEval(inputEval: Any, replaceEval: Any, posEval: Any, lenEval: Any): Any = {
+    val inputStr = inputEval.asInstanceOf[UTF8String]
+    val replaceStr = replaceEval.asInstanceOf[UTF8String]
+    val position = posEval.asInstanceOf[Int]
+    val length = lenEval.asInstanceOf[Int]
+    Overlay.calculate(inputStr, replaceStr, position, length)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    defineCodeGen(ctx, ev, (input, replace, pos, len) => "org.apache.spark.sql.catalyst." +
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: nit: style. we need a line break after `=>` in this case.

**Comment by ueshin** [comment]
@beliefer Thanks for checking.
Then I'm fine with the current behavior.
LGTM, except for the two nits.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 10]
```diff
@@ -454,6 +454,66 @@ case class StringReplace(srcExpr: Expression, searchExpr: Expression, replaceExp
...
+  override def nullSafeEval(inputEval: Any, replaceEval: Any, posEval: Any, lenEval: Any): Any = {
+    val inputStr = inputEval.asInstanceOf[UTF8String]
+    val replaceStr = replaceEval.asInstanceOf[UTF8String]
+    val position = posEval.asInstanceOf[Int]
+    val length = lenEval.asInstanceOf[Int]
+    Overlay.calcuate(inputStr, replaceStr, position, length)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    nullSafeCodeGen(ctx, ev, (input, replace, pos, len) => {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Yes, changed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 8 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: OK. I learned it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 9 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: OK. I changed it.

**Comment by beliefer** [comment]
@ueshin Thanks for your review. I learned more.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Comment by beliefer** [comment]
@ueshin Thanks for your help. @gatorsmile @wangyum @dongjoon-hyun Thanks for all your review.

**Pull request closed without merging**